### ISSUE TITLE
Feature Request: parallel_call_once/parallel_once_flag

### ISSUE THREAD

**Issue by BenFrantzDale** [oneapi-src/oneTBB#267]
From https://software.intel.com/en-us/node/841360 (`isolated_task_group`) there's an example `class parallel_once` which acts a bit like `std::call_once` https://en.cppreference.com/w/cpp/thread/call_once, but rather than blocking, follow-on threads can help get the work done. Prior to finding that example code, I was thinking about the same idea, but I think the better way to do it would be to follow the standard's API and provide a `tbb::once_flag` and `tbb::parallel_call_once` (or maybe just `tbb::call_once`?)

The example on https://software.intel.com/en-us/node/841360 seems very close; it just lacks proper exception handling, since `std::call_once` doesn't count exceptions as successful calls. In a correct parallel version, the initial calling thread would throw, but then a waiting (helping) thread would get to try calling its function while any other waiting threads would re-try helping until all calls either throw or one succeeds.

The other change I'd consider vs the example is to potentially have the `tbb::isolated_task_group` be an `std::optional< tbb::isolated_task_group>` (or have it heap-allocated) so the `tbb::isolated_task_group` only has to exist while in use. (I'm not clear on the cost in memory and/or other resources of keeping potentially thousands of `tbb::isolated_task_group` instances around. For reference, for at least one implementation, `sizeof(std::once_flag) == 4` https://godbolt.org/z/WhcesE)

If this is of interest, I could help implement it.

**Comment by alexey-katranov** [comment]
The idea seems interesting.

As for isolated_task_group, the main problem how to delete it (the complete flag can be set earlier than all thread leaves the `task_group`). I have a complicated idea with 4 state for once_flag and additional reference count (barrier) for helping threads. I will share the details later if there is no better solution. 

Another issue is user threads, each user thread works in its own `arena`, so it is not so easy to join a user thread to help. Perhaps, attached `task_arena` can be used (but it is also complicated solution how to transfer this information to the helping threads).

So, if you are not afraid of possible complexities, feel free to open full request with your contribution (use onetbb_2021 branch for the baseline). And/or we can discuss in details how to overcome the issues.

**Comment by BenFrantzDale** [comment]
Here's what I've got: https://godbolt.org/z/qTdET3

**Comment by alexey-katranov** [comment]
The approach seems reasonable. I'd suggest using attached `task_arena`. I.e. `task_arena` can be initialized on the first thread with the `attach` argument. It can reduce some overheads when the calling thread already initialized the implicit `arena`.

My crazy idea is to have `once_flag` of a pointer size. It can be pointer with 4 states: "uninitialized" (`U`), "locked" (`L`), "real_pointer" (`P`), "initialized" (`I`). The state machine:
```
U-->L<-->P-->I
```
The first thread algorithm:
1. `U-->P`
1. Call function
1. `P-->I`
1. Wait for the moonlighting thread completion

The moonlighting threads algorithm:
1. `P-->L`
1. Increase reference of the moonlighting threads
1. `L-->P`
1. Moonlight
1. Decrease the reference

In addition, we can avoid dynamic memory and allocate the state structure on the stack of the first thread.

**Comment by BenFrantzDale** [comment]
From https://software.intel.com/en-us/node/506359 it wasn't clear to me if `task_arena(task_arena::attach)` does what we'd want. In particular, it sounds like it makes a `task_arena` that's essentially an alias to the arena the calling thread is already in. Does that mean that a subsequent call to `arena.execute(f)` would still be allowed to moonlight above that call to `execute` so long as it's within the attached arena? As in:
```
task_arena a;
a.execute([]() {
    task_group g;
    g.run(foo);
    task_arena a1(task_arena::attach);
    a1.execute([]() {
        task_group g1;
        g1.run_and_wait(bar); // Can this thread moonlight in calling foo since a1 and a are "the same"?
    });
    g.wait();
});
```

**Comment by BenFrantzDale** [comment]
For your one-pointer-size suggestion, I'm assuming your picturing something like a `std::atomic<void*>` where `U` is `nullptr`, `L` and `P` are some special hopefully-not-real-memory value like `0x1` and `0x2` (or `size_t(-1)` and `size_t(-2)`?)

The first-thread's-stack storage is interesting. I used `std::shared_ptr` to let moonlighting threads `wait()` even after the first thread has left. You are suggesting instead that the first thread could spin waiting for the moonlighters to clear out. That seems reasonable. (FWIW, `sizeof(std::once_flag) == 4` vs `sizeof(void*) == 8`, so storing a pointer seems totally reasonable.

**Comment by alexey-katranov** [comment]
As for nested arena execute, it will work almost like there is no nested execute (there might be small differences in enqueue tasks and priorities in flow graph but it is not really important). However, `g1` is better to be `isolated_task_group` to prevent the moonlighting thread against taking tasks from other `task_group`s, e.g. `g`.

As for one-pointer approach, yes it can be `std::atomic<void*>` with special values. (Perhaps, I'd use `std::atomic<std::uintptr_t>` because conversion from pointer to `std::uintptr_t` and vice verse is defined better than conversion of some integers to the pointers.)

**Comment by BenFrantzDale** [comment]
I think you are saying my understanding is correct that the thread running `g1.run_and_wait(bar);` could wind up being the thread to call `foo`. If so, I think `task_arena::attach` is inadequate for a `call_once` because if that thread is allowed to take tasks above the `call_once` stack frame, then it could itself wind up back in `call_once` leading to unbounded recursion. As in
```
parallel_once_flag flag;
std::optional<float> result;
task_arena a;
a.execute([&]() {
    task_group g;
    g.run([&]() {
        doWork0();
        call_once(flag, [&]() { result = foo(); });
        doMoreWork0(*result);
    });
    g.run([&]() {
        doWork1();
        call_once(flag, [&]() { result = foo(); });
        doMoreWork1(*result);
    });
    g.wait();
});
```
If a thread calling `call_once` is allowed to take the other `g.run` task, then it could itself wind up in the other `call_once`. Is that a valid concern for `task_arena::attach`? That said, it does sound like `isolated_task_group` is what I'm looking for. (I'm just trying to be sure I grok `task_arena`.)


As for the pointer approach, I agree about `std::uintptr_t`. It seems dicy to assume that `0x1` and `0x2` aren't in valid memory, although on my desktop computer, it's almost certainly true. I guess one saving grace would be to `static_assert(alignof(ControlBlock) > 2)` which would mean that the pointer to the `ControlBlock` can't possibly be `0x1` or `0x2` since those two bits must always be zero for a valid (aligned) pointer. That said, a second field, bringing the `once_flag` size up to two pointers' size seems reasonable too.

**Comment by BenFrantzDale** [comment]
FWIW, I've now used this to great effect. I'm increasingly convinced that `do_once` is a very powerful fundamental building block. In particular, it makes it much easier to write a correct non-deadlocking cache where the work can be done in parallel.

**Comment by BenFrantzDale** [comment]
@alexey-katranov, I've done more with this idea and am increasingly finding it to be very powerful. (I also have other interesting things built on it.) Would you consider it for inclusion in tbb? I can prepare a PR if you are interested. We could do a video call if you'd like to discuss in detail.

**Comment by alexey-katranov** [comment]
@BenFrantzDale , sorry for the delay with the response. I completely agree that it is really useful pattern. I really appreciate if you are able to open merge request. If you have questions, we can have a call to discuss them before MR.

**Comment by BenFrantzDale** [comment]
@alexey-katranov I'd like to talk first. I don't see a way to DM on github. Can you email me at [github user name]@gmail.com?

**Comment by BenFrantzDale** [comment]
@aleksei-fedotov this can be closed, right?
